### ISSUE TITLE
Interest check: Integration between nvim-autopairs and LuaSnip

### ISSUE THREAD

**Issue by dajohans** [windwp/nvim-autopairs#83]
I'm on a quest to obtain automatic pair completion in (Neo)Vim which has an equal level of polish as that in many IDEs. Personally, I think Eclipse probably has the best auto pair behaviour that I've experienced. Naturally, I wish this to change so that instead (Neo)Vim has the best auto pair behaviour that I've experienced.

I've had a quite fruitful discussion in https://github.com/L3MON4D3/LuaSnip/issues/77 and there I think we've come to the conclusion that it is probably quite easy to make nvim-autopairs perform pair completion using snippets in LuaSnip. See https://github.com/L3MON4D3/LuaSnip/issues/77#issuecomment-881988789 where I obtain pretty good result by changing just 3 lines of code in nvim-autopairs. But if you agree to accept this feature, I suppose we should add some more code than these 3 lines in order to make it an optional feature.

What do you think, @windwp ? Could this be a feature you're willing to accept into nvim-autopairs? I'd be willing to work on this, so I'm not necessarily asking you or anybody else to do the work for me.

For awareness of this issue: Tagging @L3MON4D3 the author of LuaSnip.

### What would be the advantages of this?
I give two arguments in https://github.com/L3MON4D3/LuaSnip/issues/77 and I repeat them below:
1.  Having a more _consistent_ experience with pair completion when using both a pair completion plugin and a snippet plugin at the same time. If the pair `()` comes from a snippet then I can jump around as one can usually do in snippets but if the pair comes from an autopair plugin then this can usually not be done. So by having nvim-autopairs do the actual completion with a snippet means that _all_ pairs come from a snippet and I can _always_  jump around pairs in the snippet way. So the behaviour with pairs becomes more consistent.
2. Having a more _convenient_ experience. If I've created a the pair `(|)` and `|` is my cursor and I want to leave the pair then I can press `)` to end up with `()|`. On my nordic keyboad, `)` is `shift+9` which is quite awkward. If the pair was instead created with a snippet then I could leave the pair by pressing `<tab>`, which is my keybind for jumping in snippets. Also, by leaving the pair with a snippet jump then I can leave the pair from _anywhere_ within the pair, something which I don't think is currently supported by nvim-autopairs.

I think the behaviour that I seek from automatic pair completion can be obtained by extending both nivm-autopairs and LuaSnip independently from each other. However, doing so surely results in duplication of work. To nvim-autopairs we would need to allow to jump in and out of pairs, something that LuaSnip already can do. To LuaSnip we would have to add the logic for when to complete a pair and when not to complete a pair, something which nvim-autopairs already can do.

**Comment by windwp** [comment]
I read the whole message but I only see you want to move around the pair with  `<tab>` key.

``` lua
inoremap({
    '<tab>',
    function()
        if vim.fn.pumvisible() == 1 then
            return t "<C-n>"
        elseif luasnip.expand_or_jumpable(1) then
            luasnip.expand_or_jump(1)
            return  ""
        else
            --- check the next character is a pair )"'}] then move right
            --- you need do some logic to check
            return keymap.t"<right>"
        end
    end,
    expr = true,
    noremap = true,
})

```

**Comment by dajohans** [comment]
I think your suggestion  is pretty close but not exactly what I would consider the ideal solution. I admit I didn't even think about adding my own logic for checking the next character, because I've never done that low level Vim configuration and because nvim-autopairs already could do it. But if I do it this way, then I could make `<tab>` leave a pair if the pointer is to the left of a pair ending. What LuaSnip would be able to do is to leave the pair from _anywhere_ within the pair, which can be quite convenient.

But I understand that you may not be that interested in integrating LuaSnip then.

But while I'm at it, I'll then mention some things that I would find useful in nvim-autopairs in regards to the kind of conditional keybinds in you reply. As you know, snippet plugins have some form of `jumpable()` function which can be used in conditional binds. I think it would be nice to have something like that in nvim-autopairs. This could check if the next character is any pair ending. Or maybe there could also be something similar but to check if you are currently within a pair or if you have ever left the current pair. I think that being able to see if I have ever left a pair would be super useful for conditional keybinds like this. But I'm not sure how difficult this would be to add. Usually, I think I would only want to leave a pair by pressing `<tab>` on the first time I enter the pair.

**Comment by windwp** [comment]
Firstly  I will integrate with luasnip if you have a good reason. but you already move wrong direction.
I already take a look at that code you give in  https://github.com/L3MON4D3/LuaSnip/issues/77. \

you want to remap all pairs key ` " ' (){}[]` and check to expand snippet on that key but seriously I never see any snippet finish with something like ` " ' (){}[]`.

autopairs can map to another key not only a pairs keys but It is configured by a user and the default is only map on pairs keys

if luasnip wants to auto-trigger on that key and it collapses with autopairs then I will make an option to check that.

but you want to move to the next pair by press <tab> and I don't see any reason why it needs to map all that key above to do that.

sorry if I wrong but you need to give an example situation you want to do to integrate with lua snip.

**Comment by dajohans** [comment]
> you want to remap all pairs key ` " ' (){}[]` and check to expand snippet on that key but seriously I never see any snippet finish with something like ` " ' (){}[]`.

I agree that it is really unusual to find `()` in a snippet plugin. But making the snippet expand the pair `()` allows to get the bahaviour that I want, namely, leave the pair from _anywhere_ within the pair by pressing `<tab>` and not just leave the pair from _right before_ the closing pair symbol.

This all comes down to having a _consistent_ experience with _all_ pairs in Vim. When I have both a snippet plugin and an automatic pair completion plugin, pairs can come from two sources: snippet and automatic pair completion. With most combinations of snippet and automatic pair completion plugins, you will have _different_ behaviour with pairs that come from _different_ sources. In my opinion, this makes the overall Vim experience feel really unpolished. But by using the logic for when to complete a pair which is already implemented in nvim-autopairs and combining that with the completion of a snippet plugin, then _all_ pairs would come from the _same_ completion source and that would _automatically_ give a consistent experience with all pairs.

This kind of consistency is common in IDEs and it is something I really miss since switching to Vim(and Emacs, I can never really make a final decision).

All I want in the end is a really polished and consistent experience with pairs in Vim. I thought I saw one way to obtain that by integrating LuaSnip into nvim-autopairs. But if it can be sorted out without an integration between plugins then I'd be equally happy as well.

I want two thing regarding leaving the pair:

1.  leave from _anywhere_ within the pair by pressing some key, `<tab>` in my case
2.  only leave the pair by pressing `<tab>` the _first_ time that you leave the pair.

Number 2 might sound strange so let me explain. If I have already written a line of code that contain a pair and I later return to edit that line of code inside the pair, then I probably only want to edit that specific place and then I'm going to go to some other line of code. So when I _return_ to edit an _existing_ pair, I don't want to leave the pair upon pressing `<tab>`. If I press `<tab>` in this case, I want to insert an actual tab-character. But when I write a pair and enter it for the _first_ time then it's quite likely that I'm going to want to keep writing immediately after that pair, hence I want to be able to leave the pair from _anywhere_ within the pair by a single key press, in my case `<tab>`.

Number 2 would be obtained by integrating with a snippet plugin, because once you leave the snippet you will not trigger its saved jump locations by pressing `<tab>` again. So by integrating nvim-autopairs with a snippet plugin you also automatically get the behaviour of leaving the pair by a single key press only the first time you leave the pair.

> but you want to move to the next pair by press `<tab>` key and I don't see any reason why it needs to map all that key above to do that.

I'm not sure what you mean by the second half of the sentence but I disagree with the first half. This is not what I'm trying to say. What I'm trying to say is that I want to jump the _end_ of the current pair by pressing `<tab>`. I am _not_ asking specifically for the ability to jump to the _next_ pair, although this ability would be given by integrating with a snippet plugin.

**Comment by windwp** [comment]
hi i already say it you want a tab very smart not you

**Comment by dajohans** [comment]
> you want a tab very smart not a pairs with snippet

I disagree. I don't understand what you mean by "tab very smart" and that I don't want "pairs with snippet" when I think I quite clearly argue for "pairs with snippet".

> you want to fix an issue of tab key

I don't think I agree. I don't think there's an issue with the tab key. What am I saying that makes you think there's an issue with the tab key? And what do you think this issue is?

> then you maping another pairs key with something to check snippet

I don't understand what you mean by "mapping another pairs key", inparticular the use of _another_. Which is this other pair key that you think I want to map? I also don't understand what the "check snippet" means.

I'm a bit confused. It seems that we're not understanding each other.

**Comment by windwp** [comment]
sorry for my English skill.

but  I understand you want to combine a feature of autopairs (delete, move right ... ) to luasnip with a  tab key to move around but it is a not  good idea to remap pairs key ` " ' (){}[] ` like this.

``` lua

local function t(str)
	return vim.api.nvim_replace_termcodes(str, true, true, true)
end
local parser = require('nvim-treesitter.parsers')
local _, ts_utils = pcall(require, 'nvim-treesitter.ts_utils')
function expand_parens()
	local cursor_node = ts_utils.get_node_at_cursor()
	if cursor_node ~= nil then
		print(cursor_node:type()) --this shows up in :messages
		if cursor_node:type() ~= 'source' then
			return t'(<plug>luasnip-expand-or-jump'
		end
	end
	return '('
end
vim.api.nvim_set_keymap('i', '(', 'v:lua.expand_parens()', {silent = true, expr = true, noremap = false})


  for _, rule in pairs(M.state.rules) do
        if rule.key_map ~= nil then
            if rule.is_regex == false then
                if rule.key_map == "" then
                    rule.key_map = rule.start_pair:sub((#rule.start_pair))
                end
                local key = string.format('"%s"', rule.key_map)
                if rule.key_map == '"' then key = [['"']] end
                local mapping = string.format("v:lua.MPairs.autopairs_map(%d,%s)", bufnr, key)
                --api.nvim_buf_set_keymap(bufnr, "i", rule.key_map, mapping, {expr = true, noremap = true})
                api.nvim_buf_set_keymap(bufnr, "i", rule.key_map, string.format("v:lua.expand_parens(%s)", key) , {expr = true, noremap = false})

                local key_end = rule.end_pair:sub(1,1)
                if #key_end == 1 and key_end ~= rule.key_map and rule.move_cond ~= nil then
                    mapping = string.format([[v:lua.MPairs.autopairs_map(%d, '%s')]], bufnr,key_end )
                    --vim.api.nvim_buf_set_keymap(bufnr, 'i',key_end, mapping, {expr = true, noremap = true})
                    vim.api.nvim_buf_set_keymap(bufnr, 'i',key_end, string.format("v:lua.expand_parens(%s)", key), {expr = true, noremap = false})
                end
            else
                if rule.key_map ~= "" then
                    local mapping = string.format("v:lua.MPairs.autopairs_map(%d,'%s')", bufnr, rule.key_map)
                    --api.nvim_buf_set_keymap(bufnr, "i", rule.key_map, mapping, {expr = true, noremap = true})
                    api.nvim_buf_set_keymap(bufnr, "i", rule.key_map, string.format("v:lua.expand_parens(%s)", key), {expr = true, noremap = false})
                elseif rule.is_endwise == false then
                    enable_insert_auto = true
                end
            end
        end
    end
```

you should fix that problem on `<tab>` key. I don't think it is good when you need to check a snippet to expand on that key.

**Comment by dajohans** [comment]
Am I correct if I say that you think that moving me out of a pair is not the responsibility of nvim-autopairs? And do you think that moving me out of the pair should be the responsibility of the logic behind my keybind for the (in my case) `<tab>` key?

If your answer is yes, then the source of our confusing might be found. Personally, I view it as the responsibility of the plugin for automatic pair completion to also provide the ability to move out of the completed pairs. This is why I think it is a good idea to complete the pairs with a snippet, because then the ability to move out of the pair can be built into the pair completion.

**Comment by windwp** [comment]
yes 
I already say it is not good to checking trigger a snippet on pairs key. 
if you want my plugin to trigger a luasnip. you can think about mapping it like this
``` lua
npairs.get_rule("(") :replace_endpair(function()
    return '<cmd>lua require"luasnip".expand_or_jump()<Cr>'
end)
```

here is an auto trigger when you press fold
``` json
  "fold": {
    "prefix": "fold",
    "body": "-- $1 {{{\n\t $2 \n-- }}}",
    "description": "fold for lua"
  },
```
``` lua
npairs.add_rules({
    Rule("fold", "", "lua")
    :replace_endpair(function()
        return '<cmd>lua require"luasnip".expand_or_jump()<Cr>'
    end)
})

```

**Comment by dajohans** [comment]
> if you want my plugin to trigger a luasnip. you can think about mapping it like this
> 
> ```lua
> npairs.get_rule("(") :replace_endpair(function()
>     return '<cmd>lua require"luasnip".expand_or_jump()<Cr>'
> end)
> ```

This seems to work perfectly! I honestly think this is more or less exactly what I was looking for, but I didn't know about `replace_endpair()` function and then my proposed solution became a lot more invasive. So I think this might be exactly the integration between nvim-autopairs and LuaSnip that I was looking for. I just didn't expect the integration to be possible without changes to the source code.

Thanks!

**Comment by windwp** [comment]
even this solution above is working but i still prefer to my first solution to mapping move right on tab key.

you don't need to tell lua snip check expand snippet when it just return a key ).

 when you do that you can't use dot to repeat.

**Comment by VolinM** [comment]
> I read the whole message but I only see you want to move around the pair with `<tab>` key.
> 
> ```lua
> inoremap({
>     '<tab>',
>     function()
>         if vim.fn.pumvisible() == 1 then
>             return t "<C-n>"
>          elseif luasnip.jumpable() then
>             luasnip.jump(1)
>             return  ""
>         else
>             --- check the next character is a pair )"'}] then move right
>             --- you need do some logic to check
>             return keymap.t"<right>"
>         end
>     end,
>     expr = true,
>     noremap = true,
> })
> ```

Hey, 
I'm trying to implement this, but im getting a bunch of errors. Could you explain which file and which locals you defined for this? Thanks!

**Comment by michaelfortunato** [comment]
> yes I already say it is not good to checking trigger a snippet on pairs key. if you want my plugin to trigger a luasnip. you can think about mapping it like this
> 
> npairs.get_rule("(") :replace_endpair(function()
>     return '<cmd>lua require"luasnip".expand_or_jump()<Cr>'
> end)
> 
> here is an auto trigger when you press fold
> 
>   "fold": {
>     "prefix": "fold",
>     "body": "-- $1 {{{\n\t $2 \n-- }}}",
>     "description": "fold for lua"
>   },
> 
> npairs.add_rules({
>     Rule("fold", "", "lua")
>     :replace_endpair(function()
>         return '<cmd>lua require"luasnip".expand_or_jump()<Cr>'
>     end)
> })

I am interested in this integration as well. Could some elaborate on it? 
Let `|` represent the cursor and `$<number>` denote a luasnipet node, all other pre-formated characters are literal. What I want to have it 
`(|` expand to `($1)`
`(<CR>)` expand to `(<CR><Tab>$1<CR>)`.

Essentially I just want to use nvim-autopairs for the carriage return functionality (and if possible, the pair deletion ability), and let luasnippet take care of the rest.
